processor/intracesampler/processor.go

```diff
@@ -0,0 +1,224 @@
+	config             Config
+	scaledSamplingRate uint32
+	hashSeedBytes      []byte
+}
+
+func newInTraceSamplerSpansProcessor(ctx context.Context, set processor.CreateSettings, cfg *Config, nextConsumer consumer.Traces) (processor.Traces, error) {
+
+	its := &inTraceSamplerProcessor{
+		logger:             set.Logger,
+		config:             *cfg,
+		scaledSamplingRate: uint32(cfg.SamplingPercentage * percentageScaleFactor),
+		hashSeedBytes:      i32tob(cfg.HashSeed),
+	}
+
+	return processorhelper.NewTracesProcessor(
+		ctx,
+		set,
+		cfg,
+		nextConsumer,
+		its.processTraces,
+		processorhelper.WithCapabilities(consumer.Capabilities{MutatesData: true}))
+}
+
+type FullSpan struct {
+	resource pcommon.Resource
+	scope    pcommon.InstrumentationScope
+	span     ptrace.Span
+}
+
+type TraceTreeData struct {
+
+	// map each span id to a full span object with scope and resource
+	fullSpans map[pcommon.SpanID]FullSpan
+
+	// map each span id to its children ids
+	// this enables fast leaf detection and traversal of the tree from roots
+	children map[pcommon.SpanID][]pcommon.SpanID
+
+	// spans that have no parent or their parent is not in the trace
+	roots []pcommon.SpanID
+}
+
+// this map enable us to find the parent span of a span in O(1), and all the children of a givin span
+// it also generates a span object that contains the resource and scope all at once
+// this is useful for the sampler to be able to make decisions on spans
+func spansToTraceTree(td ptrace.Traces) TraceTreeData {
+	fullSpans := make(map[pcommon.SpanID]FullSpan)
+	spanChildren := make(map[pcommon.SpanID][]pcommon.SpanID)
+
+	rss := td.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		resource := rs.Resource()
+		scopespans := rs.ScopeSpans()
+		for j := 0; j < scopespans.Len(); j++ {
+			ss := scopespans.At(j)
+			spans := ss.Spans()
+			scope := ss.Scope()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+
+				fullspan := FullSpan{
+					resource: resource,
+					scope:    scope,
+					span:     span,
+				}
+
+				fullSpans[span.SpanID()] = fullspan
+				spanChildren[span.ParentSpanID()] = append(spanChildren[span.ParentSpanID()], span.SpanID())
+			}
+		}
+	}
+
+	// find roots
+	roots := make([]pcommon.SpanID, 0)
+	for _, fullspan := range fullSpans {
+		parentSpanID := fullspan.span.ParentSpanID()
+		if _, ok := fullSpans[parentSpanID]; !ok {
+			currentSpanID := fullspan.span.SpanID()
+			roots = append(roots, currentSpanID)
+		}
+	}
+
+	traceTreeData := TraceTreeData{
+		fullSpans: fullSpans,
+		children:  spanChildren,
+		roots:     roots,
+	}
+
+	return traceTreeData
+}
+
+// check if all spans in td are from the span trace id.
+// this indicates that the processor is run after another processor
+// that emits completed traces after timeout
+// if a single trace id is found, it is returend, otherwise nil is returned
+func getSingleTraceID(td ptrace.Traces) *pcommon.TraceID {
+	var traceID *pcommon.TraceID
```

Comment:
Please avoid using using pointers to `pdata` types since it impacts garbage collection.

Reply:
No problem, but just so I understand what I am doing and applying a correct fix - is it a problem specifically with `pdata` or is it bad practice to use pointers in general?

Can you please point me to some resources that explain what is the idiomatic garbage collection-friendly alternative?

Reply:
Only for pdata types in the general case, the reason being is that pdata is an optimised layer above the raw proto types that are allocated on the heap. 

So potentially having the reference type also on the heap does mean there is some performance impacted on look up, (The pdata wrappers themselves are tiny so they aren't an issue on the stack) but it also makes it harder for the GC to know when to cleanup the types due to the indirections going on.

Reply:
I removed the reference as requested.
Please note that now to indicate "empty" trace id, I use `pcommon.NewTraceIDEmpty()`. I think it means that instead of comparing a pointer to nil, we now have to compare two `[16]byte` arrays to each other in each iteration which I guess also comes with a performance impact